syntax = "proto3";

package optimizer_agent;

service OptimizerAgent {
  // Trigger manual de otimização
  rpc TriggerOptimization(TriggerOptimizationRequest) returns (TriggerOptimizationResponse);

  // Obter status de otimização
  rpc GetOptimizationStatus(GetOptimizationStatusRequest) returns (GetOptimizationStatusResponse);

  // Listar otimizações
  rpc ListOptimizations(ListOptimizationsRequest) returns (ListOptimizationsResponse);

  // Rollback de otimização
  rpc RollbackOptimization(RollbackOptimizationRequest) returns (RollbackOptimizationResponse);

  // Obter estatísticas
  rpc GetStatistics(GetStatisticsRequest) returns (GetStatisticsResponse);

  // Health check
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);

  // ML Predictive Scheduling RPCs
  // Obter previsão de carga
  rpc GetLoadForecast(LoadForecastRequest) returns (LoadForecastResponse);

  // Obter recomendação de agendamento
  rpc GetSchedulingRecommendation(SchedulingRecommendationRequest) returns (SchedulingRecommendationResponse);

  // Obter métricas de política de agendamento
  rpc GetSchedulingMetrics(SchedulingMetricsRequest) returns (SchedulingMetricsResponse);
}

message TriggerOptimizationRequest {
  string component = 1;
  string optimization_type = 2; // WEIGHT_RECALIBRATION, SLO_ADJUSTMENT, etc.
  map<string, string> context = 3;
}

message TriggerOptimizationResponse {
  string experiment_id = 1;
  string status = 2;
  string message = 3;
}

message GetOptimizationStatusRequest {
  string optimization_id = 1;
}

message GetOptimizationStatusResponse {
  string optimization_id = 1;
  string status = 2;
  double improvement_percentage = 3;
  map<string, double> metrics = 4;
}

message ListOptimizationsRequest {
  string component = 1;
  string optimization_type = 2;
  int32 limit = 3;
  int32 skip = 4;
}

message ListOptimizationsResponse {
  repeated OptimizationSummary optimizations = 1;
  int32 total = 2;
}

message OptimizationSummary {
  string optimization_id = 1;
  string optimization_type = 2;
  string component = 3;
  double improvement_percentage = 4;
  string status = 5;
  int64 applied_at = 6;
}

message RollbackOptimizationRequest {
  string optimization_id = 1;
  string reason = 2;
}

message RollbackOptimizationResponse {
  string status = 1;
  string message = 2;
}

message GetStatisticsRequest {
  int32 days = 1;
}

message GetStatisticsResponse {
  int32 total_optimizations = 1;
  double success_rate = 2;
  double average_improvement = 3;
  map<string, int32> by_type = 4;
  map<string, int32> by_component = 5;
}

message HealthCheckRequest {}

message HealthCheckResponse {
  string status = 1;
  string version = 2;
}

// ML Predictive Scheduling Messages

message LoadForecastRequest {
  int32 horizon_minutes = 1;
  bool include_confidence_intervals = 2;
}

message LoadForecastResponse {
  repeated ForecastPoint forecast = 1;
  ForecastMetadata metadata = 2;
}

message ForecastPoint {
  string timestamp = 1;
  int32 ticket_count = 2;
  ResourceDemand resource_demand = 3;
  int32 confidence_lower = 4;
  int32 confidence_upper = 5;
}

message ResourceDemand {
  double cpu_cores = 1;
  int32 memory_mb = 2;
}

message ForecastMetadata {
  int32 model_horizon = 1;
  int32 horizon_requested = 2;
  string forecast_generated_at = 3;
  int32 data_points_used = 4;
  double confidence_level = 5;
}

message SchedulingRecommendationRequest {
  int32 current_load = 1;
  double worker_utilization = 2;
  int32 queue_depth = 3;
  double sla_compliance = 4;
}

message SchedulingRecommendationResponse {
  string action = 1;
  string justification = 2;
  double expected_improvement = 3;
  double risk_score = 4;
  double confidence = 5;
}

message SchedulingMetricsRequest {
  int32 time_range_hours = 1;
}

message SchedulingMetricsResponse {
  double average_reward = 1;
  double policy_success_rate = 2;
  map<string, int32> action_counts = 3;
  int32 states_explored = 4;
}
